### ISSUE TITLE
[Feature] Finish running a test file before running the next file in parallel mode

### ISSUE THREAD

**Issue by younesjallouli** [microsoft/playwright#27740]
Hi team,

I want to start by thanking you for all the work you've poured into creating and maintaining Playwright. It's really an amazing tool to work with on a daily basis.

### Context:
In order to test our SaaS app, we have multiple tests across multiple files, and we run these tests in parallel on multiple workers. However, achieving full isolation between files is a pipe dream for most of us here, especially in E2E testing, where you almost certainly need to create some objects in previous tests, before moving on to the next tests (like testing registering a payment method, before testing the creation of a payment, before testing a refund of that payment)

Today, you can run Playwright tests in parallel. In order to achieve that, it runs several worker processes that run at the same time. By default, test files are run in parallel. Tests in a single file are run in order, in the same worker process. The second option is to configure the entire test run to concurrently execute all tests in all files using the option `fullyParallel=true`.

For both values of fullyParallel, Playwright will max out the utilization of workers and in the case of `workers=4` for example, the execution model will look like this : 

![image](https://github.com/microsoft/playwright/assets/116430697/5dc74826-4b26-4e2f-bc40-a690acea3096)

### The missing part in Playwright:

Unfortunately, if we want to finish running a test file (all payment methods creation in the previous example) before moving on to the next files, we need a way to say to Playwright that we want to run test files one by one sequentially, but inside test files, everything should run in parallel. The tradeoff would be execution speed since workers will not be as utilized as in the previous models, **however** code duplication between test files would be minimal, as each test file would tackle a specific domain, and test dependencies would not be an issue anymore.

**So this model is definitely worth it as the company test suite matures and becomes bigger.**

![image](https://github.com/microsoft/playwright/assets/116430697/2b65b870-7328-4de9-972c-a8490cd6e81c)

Thank you in advance

**Comment by marcusNumminen** [comment]
Hi!
I think I had a similar problem that I managed to solve using project dependencies. I had some "destructive tests" that I needed to run separately. So what I did was to split the tests up in different projects and set a dependency between them. Something like this:
```
projects: [
    {
      name: 'api-setup',
      testMatch: /api\.setup\.ts/,
      teardown: 'api-cleanup'
    },
    {
      name: 'api-cleanup',
      testMatch: /api\.cleanup\.ts/
    },
    {
      name: 'parallel-runs',    
      testIgnore: '*.seq.*',
      testDir: './tests/
      dependencies: ['api-setup'],
    },
    {
      name: 'sequential-1',
      testMatch: 'seq1.*',
      testDir: './tests',
      dependencies: ['parallel-runs'],
    },
    {
      name: 'sequential-2',
      testMatch: 'seq2.*',
      testDir: './tests',
      dependencies: ['sequential-1'],
    },
]
```
Then when you should run the tests you run the project 'sequential-2'.

I hope this could help.
//Marcus

**Comment by younesjallouli** [comment]
Hi Marcus, thank you for your feedback,

However, this solution is not right, because in the dev phase, when you're writing tests and testing these, you need to be able to click the VS code extension to run a single test. With your solution dependencies would run each time you want to execute a single test, which should not happen.

**Comment by marcusNumminen** [comment]
To overcome that problem you could just add an another project in the conf like this
```
projects: [
    {
      name: 'api-setup',
      testMatch: /api\.setup\.ts/,
      teardown: 'api-cleanup'
    },
    {
      name: 'api-cleanup',
      testMatch: /api\.cleanup\.ts/
    },
    {
      name: 'parallel-runs',    
      testIgnore: '*.seq.*',
      testDir: './tests/
      dependencies: ['api-setup'],
    },
    {
      name: 'sequential-1',
      testMatch: 'seq1.*',
      testDir: './tests',
      dependencies: ['parallel-runs'],
    },
    {
      name: 'sequential-2',
      testMatch: 'seq2.*',
      testDir: './tests',
      dependencies: ['sequential-1'],
    },
    {
      name: 'test-VScode',
      testDir: './tests/',
      dependencies: ['api-setup'],
    }
]
```
and set this (test-VScode in the above example) as the default project in VS code

**Comment by younesjallouli** [comment]
@marcusNumminen thanks for the suggestion. Your solution works!

**Comment by yury-s** [comment]
This request is similar to https://github.com/microsoft/playwright/issues/17553.

**Comment by younesjallouli** [comment]
@marcusNumminen your workaround works but is not ideal since we have to reference the last project in the CLI : 
`npx playwright  --project=sequential-2` which is not intuitive

**Comment by marcusNumminen** [comment]
I know it's not an ideal solution and gets a bit "messy" but it gets the job done. One way to keep things a bit more clean is to define different conf files where you define the different "project setups" and use the default one to run in VS code. 
But lets hope that the work with  https://github.com/microsoft/playwright/issues/17553 will make things a bit easier.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you have additional information not present in this issue that you think will help prioritizing it, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
